### PR TITLE
[SystemZ][z/OS] Implement llvm.returnaddress for XPLINK

### PR DESCRIPTION
**Pull request by redstar**
The implementation follows the ELF implementation.

### CODE REVIEW

File: llvm/lib/Target/SystemZ/SystemZFrameLowering.h [hunk 1]
```diff
@@ -97,6 +106,11 @@ class SystemZELFFrameLowering : public SystemZFrameLowering {
     return usePackedStack(MF) ? SystemZMC::ELFCallFrameSize - 8 : 0;
   }
 
+  // Return the offset of the return address.
+  virtual int getReturnAddressOffset(MachineFunction &MF) const override {
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZFrameLowering.h:110]: We don't normally use `virtual` on the function definitions, only on the declarations.

File: llvm/lib/Target/SystemZ/SystemZFrameLowering.h [hunk 2]
```diff
@@ -97,6 +106,11 @@ class SystemZELFFrameLowering : public SystemZFrameLowering {
...
+    return (usePackedStack(MF) ? -2 : 14) * getPointerSize();
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZFrameLowering.h:111]: If this is the only place you need the pointer size, can't you just get it from `MF`?

**Review by uweigand — COMMENTED** [review]
Looks good in general, but see inline comments.

File: llvm/lib/Target/SystemZ/SystemZFrameLowering.h (hunk 2 above)
Comment (redstar) [reply llvm/lib/Target/SystemZ/SystemZFrameLowering.h:111]: That was my first approach. The expression is a bit clumsy, and it is needed twice (here, and for XPLINK), and therefore I refactored it into a new function. Then I realized that it should be used in other place, too, for example:

- https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/SystemZ/SystemZFrameLowering.cpp#L181
- https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/SystemZ/SystemZFrameLowering.cpp#L870
- https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/SystemZ/SystemZFrameLowering.cpp#L905
- https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/SystemZ/SystemZFrameLowering.cpp#L1031
- https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/SystemZ/SystemZFrameLowering.cpp#L1301

But this should be a follow-up PR.

File: llvm/lib/Target/SystemZ/SystemZFrameLowering.h (hunk 1 above)
Comment (redstar) [reply llvm/lib/Target/SystemZ/SystemZFrameLowering.h]: Changed.

**Review by uweigand — APPROVED** [review]
LGTM, thanks!

**Pull request merged**